python-lib/config.py

```diff
@@ -82,19 +82,35 @@ def get_hdfs_location(dataset_config: Mapping[AnyStr, Any], sf_stage_name: AnySt
     return f'{sf_stage_name}{path}/'
 
 
-def get_snowflake_to_hdfs_query(sf_location: AnyStr, sf_table_name: AnyStr) -> AnyStr:
+def get_snowflake_to_hdfs_query(sf_location: AnyStr, sf_table_name: AnyStr,
+                                sf_schema: List[Mapping[AnyStr, AnyStr]]) -> AnyStr:
     """
     Gets a COPY statement for copying from a Snowflake Table to an HDFS location
-    :param sf_location:
-    :param sf_table_name:
+    :param sf_location: HDFS location
+    :param sf_table_name: Snowflake table
+    :param sf_schema: Snowflake schema
     :return: a SQL COPY statement
     """
     # moving this function here isn't strictly necessary as it's not re-used, but it makes
     # things significantly easier to unit test.
 
+    # Generate SELECT clause and cast TIMESTAMP_TZ, TIMESTAMP_LTZ to TIMESTAMP_NTZ
+    # Also cast TIMESTAMP because users can override TIMESTAMP to actually mean TIMESTAMP_TZ per
+    # documentation: https://docs.snowflake.com/en/sql-reference/parameters.html#client-timestamp-type-mapping
+    # This is required as the COPY command does not support TZ and LTZ
+    timezoned_types = ['TIMESTAMPLTZ', 'TIMESTAMPTZ', 'TIMESTAMP']
+    columns = [
+        f'"{c["name"]}"'
+        + (f'::TIMESTAMP_NTZ AS "{c["name"]}"' if c['originalType'] in timezoned_types else '')
```

Comment:
After testing in production... looks like we can't guarantee there will be an `originalType` parameter.

```
[2020/07/17-08:25:16.264] [null-out-100] [INFO] [dku.utils]  - 0 rows successfully written (M892qecEoA)
[2020/07/17-08:25:16.264] [null-err-101] [INFO] [dku.utils]  - *************** Recipe code failed **************
[2020/07/17-08:25:16.265] [null-err-101] [INFO] [dku.utils]  - Begin Python stack
[2020/07/17-08:25:16.265] [null-err-101] [INFO] [dku.utils]  - Traceback (most recent call last):
[2020/07/17-08:25:16.266] [null-err-101] [INFO] [dku.utils]  -   File "/mnt/secondary/dataiku_data/jobs/H3_GEOINDEX_SF/Build_property_pre_model_hdfs_2020-07-17T08-24-57.210/compute_property_pre_model_hdfs_NP/custom-python-recipe/pyoutbPr6I17jlLvV/python-exec-wrapper.py", line 194, in <module>
[2020/07/17-08:25:16.266] [null-err-101] [INFO] [dku.utils]  -     exec(f.read())
[2020/07/17-08:25:16.266] [null-err-101] [INFO] [dku.utils]  -   File "<string>", line 33, in <module>
[2020/07/17-08:25:16.267] [null-err-101] [INFO] [dku.utils]  -   File "/mnt/secondary/dataiku_data/plugins/dev/snowflake-hdfs/python-lib/config.py", line 105, in get_snowflake_to_hdfs_query
[2020/07/17-08:25:16.267] [null-err-101] [INFO] [dku.utils]  -     for c in sf_schema
[2020/07/17-08:25:16.267] [null-err-101] [INFO] [dku.utils]  -   File "/mnt/secondary/dataiku_data/plugins/dev/snowflake-hdfs/python-lib/config.py", line 105, in <listcomp>
[2020/07/17-08:25:16.267] [null-err-101] [INFO] [dku.utils]  -     for c in sf_schema
[2020/07/17-08:25:16.267] [null-err-101] [INFO] [dku.utils]  - KeyError: 'originalType'
[2020/07/17-08:25:16.268] [null-err-101] [INFO] [dku.utils]  - End Python stack
[2020/07/17-08:25:16.268] [null-err-101] [INFO] [dku.utils]  - 2020-07-17 08:25:16,265 INFO Check if spark is available
[2020/07/17-08:25:16.268] [null-err-101] [INFO] [dku.utils]  - 2020-07-17 08:25:16,265 INFO Not stopping a spark context: No module named 'pyspark'
[2020/07/17-08:25:16.379] [FRT-98-FlowRunnable] [INFO] [dku.recipes.code.base] act.compute_property_pre_model_hdfs_NP - Error file found, trying to throw it: /mnt/secondary/dataiku_data/jobs/H3_GEOINDEX_SF/Build_property_pre_model_hdfs_2020-07-17T08-24-57.210/compute_property_pre_model_hdfs_NP/custom-python-recipe/pyoutbPr6I17jlLvV/error.json
[2020/07/17-08:25:16.380] [FRT-98-FlowRunnable] [INFO] [dku.recipes.code.base] act.compute_property_pre_model_hdfs_NP - Raw error is{"errorType":"\u003cclass \u0027KeyError\u0027\u003e","message":"originalType","detailedMessage":"At line 33: \u003cclass \u0027KeyError\u0027\u003e: originalType","stackTrace":[]}
[2020/07/17-08:25:16.380] [FRT-98-FlowRunnable] [INFO] [dku.recipes.code.base] act.compute_property_pre_model_hdfs_NP - Now err: {"errorType":"\u003cclass \u0027KeyError\u0027\u003e","message":"Error in Python process: originalType","detailedMessage":"Error in Python process: At line 33: \u003cclass \u0027KeyError\u0027\u003e: originalType","stackTrace":[]}
[2020/07/17-08:25:16.382] [FRT-98-FlowRunnable] [INFO] [dku.flow.activity] act.compute_property_pre_model_hdfs_NP - Run thread failed for activity compute_property_pre_model_hdfs_NP
com.dataiku.common.server.APIError$SerializedErrorException: Error in Python process: At line 33: <class 'KeyError'>: originalType
	at com.dataiku.dip.dataflow.exec.AbstractCodeBasedActivityRunner.handleErrorFile(AbstractCodeBasedActivityRunner.java:208)
	at com.dataiku.dip.dataflow.exec.AbstractCodeBasedActivityRunner.handleExecutionResult(AbstractCodeBasedActivityRunner.java:173)
	at com.dataiku.dip.dataflow.exec.AbstractCodeBasedActivityRunner.execute(AbstractCodeBasedActivityRunner.java:109)
	at com.dataiku.dip.dataflow.exec.AbstractPythonRecipeRunner.executeScript(AbstractPythonRecipeRunner.java:48)
	at com.dataiku.dip.recipes.customcode.CustomPythonRecipeRunner.run(CustomPythonRecipeRunner.java:71)
	at com.dataiku.dip.dataflow.jobrunner.ActivityRunner$FlowRunnableThread.run(ActivityRunner.java:380)
```

Reply:
In this particular example, `read_schema` had things like:

```json
[
 {"name": "parent_uprn", "type": "bigint"},
 {"name": "mastermap_toid", "type": "bigint"},
 {"maxLength": 50, "name": "sub_building_name", "type": "string"},
 {"maxLength": 50, "name": "building_name", "type": "string"},
 {"maxLength": 50, "name": "building_number", "type": "string"},
 {"maxLength": 250, "name": "dependent_thoroughfare", "type": "string"},
 {"maxLength": 250, "name": "thoroughfare", "type": "string"}
]
```

(Also didn't have any date columns.)

Reply:
Apparently only source datasets have 'originalType' column property. The transformed datasets have only 'type'.
So I suggest to use original approach where we extract the datatype from DB.

Reply:
Yea, probably something like:

```python
table_schema_re = re.search(r"(\"?(?P<schema>[^\"]+)\"?\.)?\"?(?P<table>[^\"]+)\"?", sf_table_name)

schema_clause = f'and c.TABLE_SCHEMA = \'{table_schema_re.group("schema")}\'' if table_schema_re.group("schema") else ''


sql = f"""
SELECT x.COLUMN_NAME || CASE WHEN x.DATA_TYPE = 'TIMESTAMP_TZ' THEN '::TIMESTAMP_NTZ AS ' || x.COLUMN_NAME ELSE '' END AS COLUMN_SELECTOR
FROM (
  select '"' || c.COLUMN_NAME || '"' AS COLUMN_NAME, c.DATA_TYPE
  from INFORMATION_SCHEMA.COLUMNS c
  where c.TABLE_NAME = '{result.group("table")}'
  {schema_clause}
) x
"""


cols = executor.query_to_df(sql)["COLUMN_SELECTOR"].tolist()
", ".join(cols)
```

Which should account for any case where `sf_table_name` is:

```
"PUBLIC"."FOO"
PUBLIC."BAR"
PUBLIC.BLAH
BLAH
"PUBLIC".BLAH
```

(You could use `sql = f"SHOW COLUMNS IN TABLE {sf_table_name}"`, but the `data_type` column is itself JSON.)

Reply:
Amended the logic to fetch columns from Snowflake